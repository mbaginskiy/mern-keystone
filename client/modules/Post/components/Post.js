import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import { FormattedMessage } from 'react-intl';
import renderHTML from 'react-render-html';
import styled from 'styled-components';

const Component = styled.div`
  margin: 20px 0;
  padding: 15px;
  border-radius: 2px;
`;

const Title = styled.h3`
  font-size: 28px;
  margin-bottom: 16px;
  font-weight: 400;
  color: ${props => props.theme.mainDark};

  & a {
    text-decoration: none;
    color: ${props => props.theme.mainDark};
  }
`;

const AuthorName = styled.div`
  font-size: 16px;
  margin-bottom: 16px;
  color: ${props => props.theme.mainLight};
`;

const Content = styled.div`
  font-size: 14px;
  margin-bottom: 8px;
  color: ${props => props.theme.main};
`;

const Actions = styled.div`

  & a {
    text-decoration: none;
    font-size: 14px;
    font-style: italic;
    color: ${props => props.theme.mainDark};

    &:hover {
      text-decoration: underline;
    }
  }
`;

const Post = (props) => {
  const { post: { slug, title, author, content }, type, onDelete } = props;
  return (
    <Component>
      <Title>
        <Link to={`/posts/${slug}`} >
          {title}
        </Link>
      </Title>
      <AuthorName>
        <FormattedMessage id="by" /> {author}
      </AuthorName>
      <Content>
        {renderHTML(type === 'brief' ? content.brief : content.extended)}
      </Content>
      <Actions>
        <a href="#delete" onClick={onDelete}><FormattedMessage id="deletePost" /></a>
      </Actions>
    </Component>
  );
};

Post.propTypes = {
  post: PropTypes.shape({
    author: PropTypes.string.isRequired,
    title: PropTypes.string.isRequired,
    content: PropTypes.shape({
      brief: PropTypes.string.isRequired,
      extended: PropTypes.string.isRequired,
    }).isRequired,
    slug: PropTypes.string.isRequired,
    id: PropTypes.string.isRequired,
  }).isRequired,
  type: PropTypes.oneOf(['brief', 'extended']).isRequired,
  onDelete: PropTypes.func,
};

Post.defaultProps = {
  onDelete: () => {},
};

export default Post;
