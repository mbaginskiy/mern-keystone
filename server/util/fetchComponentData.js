/*
Utility function to fetch required data for component to render in server side.
This was inspired from https://github.com/caljrimmer/isomorphic-redux-app/blob/73e6e7d43ccd41e2eb557a70be79cebc494ee54b/src/common/api/fetchComponentDataBeforeRender.js
*/
import promiseSequence from './promiseSequence';

export default function fetchComponentData(store, components, params) {
  const needs = components.reduce((prev, current) => {
    const currentNeeds = current.need || [];
    if (current.WrappedComponent && (current.WrappedComponent.need !== currentNeeds)) {
      currentNeeds.concat(current.WrappedComponent.need);
    }
    return currentNeeds.concat(prev);
  }, []);

  return promiseSequence(needs, need => store.dispatch(need(params, store.getState())));
}
