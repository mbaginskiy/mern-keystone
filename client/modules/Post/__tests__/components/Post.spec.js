import React from 'react';
import 'jest-styled-components';
import Post from '../../components/Post';
import { shallowWithIntl } from '../../../../util/react-intl-test-helper';

const post = {
  author: 'Admin User',
  title: 'Test Title 1',
  slug: 'test-title-1',
  content: {
    brief: 'test-brief',
    extended: 'test-extended',
  },
  id: '1',
};

const props = {
  post,
  type: 'brief',
  onDelete: () => {},
};

test('matches snapshot', () => {
  const wrapper = shallowWithIntl(
    <Post {...props} />,
  );

  expect(wrapper).toMatchSnapshot();
});

test('renders title and author properly', () => {
  const wrapper = shallowWithIntl(
    <Post {...props} />,
  );

  expect(wrapper.find('Post__Title').find('Link').prop('children'))
    .toBe(post.title);
  expect(wrapper.find('Post__AuthorName').prop('children'))
    .toEqual(expect.stringContaining(post.author));
});

test('renders extended content depending on type properly', () => {
  const wrapper = shallowWithIntl(
    <Post {...props} type="extended" />,
  );
  expect(wrapper.find('Post__Content').prop('children'))
    .toBe(post.content.extended);
});

test('renders brief content depending on type properly', () => {
  const wrapper = shallowWithIntl(
    <Post {...props} type="brief" />,
  );
  expect(wrapper.find('Post__Content').prop('children'))
    .toBe(post.content.brief);
});

test('calls onDelete', () => {
  const onDelete = jest.fn();
  const wrapper = shallowWithIntl(
    <Post {...props} onDelete={onDelete} />,
  );

  wrapper.find('Post__Actions a').first().simulate('click');
  expect(onDelete.mock.calls.length).toBe(1);
});
