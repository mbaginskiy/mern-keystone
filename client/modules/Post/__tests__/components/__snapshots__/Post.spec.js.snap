// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<Post__Component>
  <Post__Title>
    <Link
      replace={false}
      to="/posts/test-title-1"
    >
      Test Title 1
    </Link>
  </Post__Title>
  <Post__AuthorName>
    <FormattedMessage
      id="by"
      values={Object {}}
    />
     
    Admin User
  </Post__AuthorName>
  <Post__Content>
    test-brief
  </Post__Content>
  <Post__Actions>
    <a
      href="#delete"
      onClick={[Function]}
    >
      <FormattedMessage
        id="deletePost"
        values={Object {}}
      />
    </a>
  </Post__Actions>
</Post__Component>
`;
